src/librustdoc/test.rs

```diff
@@ -251,6 +253,42 @@ fn run_test(test: &str, cratename: &str, filename: &FileName, line: usize,
     let old = io::set_panic(Some(box Sink(data.clone())));
     let _bomb = Bomb(data.clone(), old.unwrap_or(box io::stdout()));
 
+    enum DirState {
+        Temp(tempfile::TempDir),
+        Perm(PathBuf),
+    }
+
+    impl DirState {
+        fn path(&self) -> &std::path::Path {
+            match self {
+                DirState::Temp(t) => t.path(),
+                DirState::Perm(p) => p.as_path(),
+            }
+        }
+    }
+
+    let target_path = if persist_executable {
+        let mut target_path: PathBuf = env::var("OUT_DIR")
+                .expect("Rustdoc must be run with Cargo to persist doctest executables.").into();
+
+        while !target_path.ends_with("target") {
```

Comment:
This doesn't seem safe.  The target directory is not necessarily named "target".  Also, cargo has a bit of an expectation of managing the directory layout.  Normally tools use something like `-o` to control their output.